### PR TITLE
[DNM]messenger: client bound OSD's public messenger

### PR DESCRIPTION
**Pull request by Aran85**
at multiple NIC host,the network has some L3 routing problems, and source bind may
prevent some of them.

Signed-off-by: Zengran Zhang zhangzengran@h3c.com

### CODE REVIEW

**Comment by Aran85** [comment]
@xiaoxichen could you review the PR? i think you also meet some L3 routing problems before. regards!

**Comment by gregsfortytwo** [comment]
I don't understand how this is supposed to fix things. Are you seeing routing issues *between* the cluster? The ms_public messenger doesn't initiate connections to anything except the monitors, and since the messenger is bound to its address I believe any outgoing connections it starts should be as well.

**Comment by Aran85** [comment]
@gregsfortytwo thanks for review. The issues is that in our ceph hosts ,there is a third NIC used for other purpose, and the configure for the NIC has a gateway. When the public network down, the outgoing connection routed to gateway and the messages to monitor successful sended.
Am I makeing sense?

**Comment by xiaoxichen** [comment]
But in this case, the OSD will be marked down by other OSDs.

**Comment by Aran85** [comment]
@xiaoxichen yes, the purpose of the PR is  prevent the osd whose public network had down reporting to monitor though  other route.

**Comment by yuyuyu101** [comment]
Hmm, without considering the detail impl, this idea makes sense to me.  Maybe we could make this default behavior?

**Comment by Aran85** [comment]
@yuyuyu101 I'm testing MSG_DONTROUTE  flag when calling ::send() at Pipe::tcp_write(). i think it also ok. what do you think? Thanks!

**Comment by Aran85** [comment]
@yuyuyu101 sorry,MSG_DONTROUTE  is not ok.because it will not prevent route when connecting,only works after connected.
I also test the SO_DONTROUTE which are one flag of setsockopt,it works well  in my simple network environment,but i think in complex environment in which case that  some osds have different network segment and depend on route(not L2 swich)  may have problem.

client bind and  SO_BINDTODEVICE may have same effect,but SO_BINDTODEVICE works only for some socket types.

if the client bind is best, i think record the address to be bound in the connection when calling get_connect()  be better, so the ip could decide by osdmap .

**Comment by yuyuyu101** [comment]
update this PR?

**Comment by Aran85** [comment]
how do you think about SO_BINDTODEVICE option? thanks! Or just try client bind?

**Comment by yuyuyu101** [comment]
I guess bind should be easy to understand

**Comment by Aran85** [comment]
thank you ! i will update the PR later.

**Comment by Aran85** [comment]
@yuyuyu101 updated, plz have a look at the impl,thank you!

File: src/msg/msg_types.h [hunk 1]
```diff
@@ -31,6 +31,7 @@ class entity_name_t {
 public:
   __u8 _type;
   int64_t _num;
+  int _ntype;	// used by connect bind
```
Comment (yuyuyu101) [inline src/msg/msg_types.h]: I suppose we don't add a field here since _ntype isn't necessary for transmitting.

Comment (Aran85) [inline src/msg/msg_types.h]: move it to next public field in the same class ?

File: src/msg/simple/Pipe.cc [hunk 2]
```diff
@@ -905,6 +906,30 @@ int Pipe::connect()
     goto fail;
   }
 
+  if (peer_ntype) {
+    entity_addr_t addr2bind;
+    switch (peer_ntype) {
+      case CEPH_ADDR_TYPE_CLIENT:
+      case CEPH_ADDR_TYPE_HBFRONT:
+	addr2bind = msgr->cct->_conf->public_addr;
```
Comment (yangruifeng) [inline src/msg/simple/Pipe.cc]: need break.

**Comment by yuyuyu101** [comment]
since I don't have confident this is ok for other normal cases(don't want to make a surprise for users), we may want to introduce more reviewers involved in....
@liewegas @athanatos

File: src/msg/simple/Pipe.cc [hunk 3]
```diff
@@ -905,6 +906,30 @@ int Pipe::connect()
...
+	addr2bind = msgr->cct->_conf->cluster_addr;
+      case CEPH_ADDR_TYPE_HBBACK:
+	addr2bind = msgr->cct->_conf->osd_heartbeat_addr;
+	if (addr2bind.is_blank_ip())
+	  addr2bind = msgr->cct->_conf->cluster_addr;
+      default:
+	assert(0 == "unkown peer ntype");
+    }
+    if (addr2bind.is_ip())
+      addr2bind.set_port(0);
```
Comment (yuyuyu101) [inline src/msg/simple/Pipe.cc]: this should be a common helper function for other messengers

**Comment by yuyuyu101** [comment]
need rebase

**Comment by Aran85** [comment]
@yuyuyu101 rebased!

**Comment by yuyuyu101** [comment]
@liewegas @gregsfortytwo except the current impl, is this idea reasonable enough to continue?

**Comment by liewegas** [comment]
This makes more sense to me now--I didn't realize that bind before connect was a supported thing.  This could also fix the issue where you specify a cluster network, we pick an apprpriate ip on the right interface, but our first connection is to a mon on a different network and we learn the wrong ip (this was reported a while ago; not sure if it is still a problem).  

Anyway, as long as we have an option to control whether we bind before connect (ms_bind_before_connect?) this seems fine.

The implemention here seems more complex than it needs to be, though.  It seems like we just need a Messenger::client_bind() that picks an ip based on public_addr or whatever, called when we do bind on the other messengers.  Then use that on any outgoing connections...

**Comment by gregsfortytwo** [comment]
What @liewegas said sounds good to me. :)

**Comment by Aran85** [comment]
@liewegas @yuyuyu101 
the hbclient messenger will connect on both networks...maybe just a  Messenger::client_bind() could not help it to choose a address for bind?

**Comment by athanatos** [comment]
I guess that's true, hbclient would need to be hbclient_front and hbclient_back?

**Comment by yuyuyu101** [comment]
@Aran85 ping?

**Comment by Aran85** [comment]
@yuyuyu101 Initiate a discussion in the dev maillist?

**Comment by yuyuyu101** [comment]
@Aran85 I think so

**Comment by liewegas** [comment]
Yeah, I think we'd need hbclient_front and hbclient_back.  Slightly annoying, but not too bad.

**Comment by yuyuyu101** [comment]
@Aran85 ping?

**Comment by dachary** [comment]
feel free to re-open once you have time to work on it

**Pull request closed without merging**